### PR TITLE
Replace arangodb::Mutex/ConditionVariable with std::mutex/condition_variable

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

Backport needed only for fix cppcheck errors for agency State

Replace arangodb::Mutex/ConditionVariable with std::mutex/condition_variable
No real difference, behavior should be same because both implementation using pthread/win-api

1. Allow us to use clang deadlock detector
2. Simplify our code
3. Make it standard and portable way
4. Allow us to try to use other mutex/condvar implementation
5. Better readability
6. More difficult to misuse (few places with old condvar)
7. Automatically improvement when update standard library

Also fix cppcheck for these files, I was forced by jenkins

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1236
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Cluster/HeartbeatThread.cpp [hunk 1]
```diff
@@ -1202,9 +1203,9 @@ void HeartbeatThread::runAgent() {
   // simple loop to post dead threads every hour, no other tasks today
   while (!isStopping()) {
     try {
-      CONDITION_LOCKER(locker, _condition);
+      std::unique_lock locker{_condition.mutex};
       if (!isStopping()) {
-        locker.wait(std::chrono::hours(1));
+        _condition.cv.wait_for(locker, std::chrono::hours{1});
```
Comment (MBkkt) [inline arangod/Cluster/HeartbeatThread.cpp:1208]: Looks really strange, why cannot we just not creating heartbeat thread on agency?
Anyway not related to this PR, just thoughts

File: arangod/Cluster/AgencyCallback.cpp [hunk 2]
```diff
@@ -203,7 +202,11 @@ bool AgencyCallback::executeByCallbackOrTimeout(double maxTimeout) {
 
     // we haven't yet been signaled. so let's wait for a signal or
     // the timeout to occur
-    if (!_cv.wait(static_cast<uint64_t>(maxTimeout * 1000000.0))) {
+    std::unique_lock lock{_cv.mutex};
+    if (_cv.cv.wait_for(lock, std::chrono::microseconds{static_cast<uint64_t>(
+                                  maxTimeout * 1000000.0)}) ==
+        std::cv_status::timeout) {
+      lock.unlock();
```
Comment (MBkkt) [inline arangod/Cluster/AgencyCallback.cpp:209]: Incorrect replacement, looks like _cv.mutex already was locked here.

So needs to write
```suggestion
    std::unique_lock lock{_cv.mutex, std::adopt_lock};
    auto const cv_status = _cv.cv.wait_for(lock, std::chrono::microseconds{static_cast<uint64_t>(
                                  maxTimeout * 1000000.0)});
    lock.release();
    if (cv_status == std::cv_status::timeout) {
```


Btw it's place, where in old code tsan detected lock order inversion.
Maybe someone should fix it?

File: arangod/Agency/Compactor.cpp [hunk 3]
```diff
@@ -47,9 +46,10 @@ void Compactor::run() {
   while (true) {
     bool falseAlarm = true;
     {
-      CONDITION_LOCKER(guard, _cv);
+      std::unique_lock guard{_cv.mutex};
       if (!_wakeupCompactor) {
-        _cv.wait(5000000);  // just in case we miss a wakeup call!
+        _cv.cv.wait_for(guard, std::chrono::seconds{5});
+        // just in case we miss a wakeup call!
```
Comment (MBkkt) [inline arangod/Agency/Compactor.cpp:52]: It's just not possible.

File: arangod/Agency/Inception.cpp [hunk 4]
```diff
@@ -547,12 +546,12 @@ void Inception::run() {
...
+  _cv.cv.notify_all();
 }
 
 // @brief Let external routines, like Agent::gossip(), signal our condition
 void Inception::signalConditionVar() {
-  CONDITION_LOCKER(guard, _cv);
-  guard.broadcast();
+  std::lock_guard guard{_cv.mutex};
+  _cv.cv.notify_all();
 }
```
Comment (MBkkt) [inline arangod/Agency/Inception.cpp:557]: Not sure but looks like signal/notify_one should be instead of broadcast/notify_all

File: arangod/Agency/Store.cpp [hunk 5]
```diff
@@ -98,8 +96,8 @@ Store::Store(arangodb::ArangodServer& server, Agent* agent,
 /// Copy assignment operator
 Store& Store::operator=(Store const& rhs) {
   if (&rhs != this) {
-    MUTEX_LOCKER(otherLock, rhs._storeLock);
-    MUTEX_LOCKER(lock, _storeLock);
+    std::lock_guard otherLock{rhs._storeLock};
+    std::lock_guard lock{_storeLock};
```
Comment (MBkkt) [inline arangod/Agency/Store.cpp:100]: scoped_lock

File: arangod/Agency/Store.cpp [hunk 6]
```diff
@@ -112,8 +110,8 @@ Store& Store::operator=(Store const& rhs) {
 /// Move assignment operator
 Store& Store::operator=(Store&& rhs) {
   if (&rhs != this) {
-    MUTEX_LOCKER(otherLock, rhs._storeLock);
-    MUTEX_LOCKER(lock, _storeLock);
+    std::lock_guard otherLock{rhs._storeLock};
+    std::lock_guard lock{_storeLock};
```
Comment (MBkkt) [inline arangod/Agency/Store.cpp:114]: scoped_lock

File: arangod/Agency/Supervision.cpp [hunk 7]
```diff
@@ -973,7 +960,7 @@ void Supervision::reportStatus(std::string const& status) {
   bool persist = false;
 
   {  // Do I have to report to agency under
-    _lock.assertLockedByCurrentThread();
+
```
Comment (MBkkt) [inline arangod/Agency/Supervision.cpp:963]:
```suggestion
```

File: arangod/Agency/Supervision.h [hunk 8]
```diff
@@ -296,11 +295,8 @@ class Supervision : public arangodb::Thread {
       std::string const& coordinatorID, uint64_t rebootID,
       bool coordinatorFound);
 
-  /// @brief Migrate chains of distributeShardsLike to depth 1
-  void fixPrototypeChain(VPackBuilder&);
```
Comment (MBkkt) [inline arangod/Agency/Supervision.h:300]: was unused and without implementation

File: arangod/Statistics/figures.h [hunk 9]
```diff
@@ -70,8 +68,8 @@ struct Distribution {
   }
 
   Distribution& operator=(Distribution& other) {
-    MUTEX_LOCKER(l1, _mutex);
-    MUTEX_LOCKER(l2, other._mutex);
+    std::lock_guard l1{_mutex};
+    std::lock_guard l2{other._mutex};
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/Statistics/figures.h:72]: Will be done in separate PR

File: arangod/Agency/Store.cpp (hunk 6 above)
Comment (MBkkt) [reply arangod/Agency/Store.cpp:114]: Will be done in separate PR

File: arangod/Agency/Store.cpp (hunk 5 above)
Comment (MBkkt) [reply arangod/Agency/Store.cpp:100]: Will be done in separate PR

File: arangod/Statistics/figures.h [hunk 10]
```diff
@@ -102,8 +100,8 @@ struct Distribution {
   }
 
   void add(Distribution& other) {
-    MUTEX_LOCKER(lock, _mutex);
-    MUTEX_LOCKER(lock2, other._mutex);
+    std::lock_guard lock{_mutex};
+    std::lock_guard lock2{other._mutex};
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/Statistics/figures.h:104]: Will be done in separate PR

File: arangod/Cluster/AgencyCallback.cpp (hunk 2 above)
Comment (MBkkt) [reply arangod/Cluster/AgencyCallback.cpp]: Suggestion applied

File: arangod/Agency/Inception.h [hunk 11]
```diff
@@ -74,8 +74,8 @@ class Inception : public Thread {
   Agent& _agent;                            //< @brief The agent
   arangodb::basics::ConditionVariable _cv;  //< @brief For proper shutdown
   std::unordered_map<std::string, size_t>
-      _acked;                        //< @brief acknowledged config version
-  mutable arangodb::Mutex _ackLock;  //< @brieg Guard _acked
+      _acked;                   //< @brief acknowledged config version
+  mutable std::mutex _ackLock;  //< @brieg Guard _acked
```
Comment (jsteemann) [inline arangod/Agency/Inception.h:78]:
```suggestion
  mutable std::mutex _ackLock;  //< @brief Guard _acked
```

File: arangod/Agency/State.cpp [hunk 12]
```diff
@@ -680,49 +677,39 @@ log_t State::atNoLock(index_t index) const {
...
 
   // If index above highest entry
-  if (_log.size() > 0 && index > _log.back().index) {
+  if (!_log.empty() && index > _log.back().index) {
     return -1;
   }
 
   // Catch exceptions and avoid overflow:
-  if (index < _cur || index - _cur > _log.size()) {
+  if (index < _cur || index - _cur >= _log.size()) {
```
Comment (jsteemann) [inline arangod/Agency/State.cpp:688]: Fix should be backported to stable versions as well.

File: arangod/Agency/State.cpp [hunk 13]
```diff
@@ -680,49 +677,39 @@ log_t State::atNoLock(index_t index) const {
...
 }
 
 /// Have log with specified index and term
 bool State::has(index_t index, term_t term) const {
-  MUTEX_LOCKER(mutexLocker, _logLock);  // Cannot be read lock (Compaction)
+  std::lock_guard mutexLocker{_logLock};  // Cannot be read lock (Compaction)
 
   // Catch exceptions and avoid overflow:
-  if (index < _cur || index - _cur > _log.size()) {
+  if (index < _cur || index - _cur >= _log.size()) {
```
Comment (jsteemann) [inline arangod/Agency/State.cpp:700]: same

File: arangod/Agency/State.cpp [hunk 14]
```diff
@@ -1247,11 +1234,11 @@ bool State::loadRemaining(index_t cind) {
 
 /// Find entry by index and term
 bool State::find(index_t prevIndex, term_t prevTerm) {
-  MUTEX_LOCKER(mutexLocker, _logLock);
-  if (prevIndex > _log.size()) {
+  std::lock_guard mutexLocker{_logLock};
+  if (prevIndex >= _log.size()) {
```
Comment (jsteemann) [inline arangod/Agency/State.cpp:1238]: this fix should be backported to stable versions.

File: arangod/Cache/Manager.cpp [hunk 15]
```diff
@@ -28,12 +28,12 @@
 #include <set>
 #include <thread>
 #include <utility>
+#include "Basics/cpu-relax.h"
```
Comment (jsteemann) [inline arangod/Cache/Manager.cpp:31]: why is this included moved to the top here and in other files?

File: arangod/V8Server/V8DealerFeature.cpp [hunk 16]
```diff
@@ -1607,14 +1608,14 @@ void V8DealerFeature::shutdownContexts() {
 
   // wait for one minute
```
Comment (jsteemann) [inline arangod/V8Server/V8DealerFeature.cpp:1609]: Comment not accurate anymore. Can be removed.
```suggestion
```

File: lib/Basics/ConditionVariable.h
Comment (jsteemann) [inline lib/Basics/ConditionVariable.h:1]: Shouldn't we remove this file as well, and replace occurrences of `arangodb::basics::ConditionVariable` with std::mutex and std::condition_variable?

**Review by jsteemann — APPROVED** [review]
LGTM, but please check review comments.

File: lib/Basics/ConditionVariable.h
Comment (MBkkt) [reply lib/Basics/ConditionVariable.h:1]: Really not convenient for replace

Comment (MBkkt) [reply lib/Basics/ConditionVariable.h:1]: Maybe in some next weekend :)

File: arangod/Agency/State.cpp (hunk 14 above)
Comment (MBkkt) [reply arangod/Agency/State.cpp:1238]: Yes, here

File: arangod/Cache/Manager.cpp (hunk 15 above)
Comment (MBkkt) [reply arangod/Cache/Manager.cpp:31]: Check last commit :( I tried to replace

https://github.com/arangodb/arangodb/pull/18448/commits/1365553b076454621d0e225eb6a7e99bb53e6889

File: arangod/V8Server/V8DealerFeature.cpp (hunk 16 above)
Comment (MBkkt) [reply arangod/V8Server/V8DealerFeature.cpp:1609]: 10 * 60 * sleep(10ms) it still true

Comment (MBkkt) [reply arangod/V8Server/V8DealerFeature.cpp:1609]: Fixed to `wait no more than one minute`

File: arangod/Replication/Syncer.cpp [hunk 17]
```diff
@@ -408,7 +406,7 @@ void Syncer::JobSynchronizer::request(std::function<void()> const& cb) {
 /// the syncer was stopped/aborted already)
 bool Syncer::JobSynchronizer::jobPosted() {
   while (true) {
-    CONDITION_LOCKER(guard, _condition);
+    std::unique_lock guard{_condition.mutex};
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/Replication/Syncer.cpp:409]: Yes, used for wait_for

File: arangod/Cache/Manager.cpp (hunk 15 above)
Comment (MBkkt) [reply arangod/Cache/Manager.cpp:31]: It wasn't intentional as you can see

Comment (MBkkt) [reply arangod/Cache/Manager.cpp]: Fixed
